src/main.rs

```diff
@@ -0,0 +1,567 @@
+				user_config,
+				chain_params,
+			);
+			(getinfo_resp.latest_blockhash, fresh_channel_manager)
+		}
+	};
+
+	// Step 10: Sync ChannelMonitors and ChannelManager to chain tip
+	let mut chain_listener_channel_monitors = Vec::new();
+	let mut cache = UnboundedCache::new();
+	let mut chain_tip: Option<poll::ValidatedBlockHeader> = None;
+	if restarting_node {
+		let mut chain_listeners =
+			vec![(channel_manager_blockhash, &mut channel_manager as &mut dyn chain::Listen)];
+
+		for (blockhash, channel_monitor) in channelmonitors.drain(..) {
+			let outpoint = channel_monitor.get_funding_txo().0;
+			chain_listener_channel_monitors.push((
+				blockhash,
+				(channel_monitor, broadcaster.clone(), fee_estimator.clone(), logger.clone()),
+				outpoint,
+			));
+		}
+
+		for monitor_listener_info in chain_listener_channel_monitors.iter_mut() {
+			chain_listeners.push((
+				monitor_listener_info.0,
+				&mut monitor_listener_info.1 as &mut dyn chain::Listen,
+			));
+		}
+		chain_tip = Some(
+			init::synchronize_listeners(
+				&mut bitcoind_client.deref(),
+				args.network,
+				&mut cache,
+				chain_listeners,
+			)
+			.await
+			.unwrap(),
+		);
+	}
+
+	// Step 11: Give ChannelMonitors to ChainMonitor
+	for item in chain_listener_channel_monitors.drain(..) {
+		let channel_monitor = item.1 .0;
+		let funding_outpoint = item.2;
+		chain_monitor.watch_channel(funding_outpoint, channel_monitor).unwrap();
+	}
+
+	// Step 13: Optional: Initialize the NetGraphMsgHandler
+	// XXX persist routing data
+	let genesis = genesis_block(args.network).header.block_hash();
+	let router =
+		Arc::new(NetGraphMsgHandler::new(genesis, None::<Arc<dyn chain::Access>>, logger.clone()));
+
+	// Step 14: Initialize the PeerManager
+	let channel_manager: Arc<ChannelManager> = Arc::new(channel_manager);
+	let mut ephemeral_bytes = [0; 32];
+	rand::thread_rng().fill_bytes(&mut ephemeral_bytes);
+	let lightning_msg_handler =
+		MessageHandler { chan_handler: channel_manager.clone(), route_handler: router.clone() };
+	let peer_manager: Arc<PeerManager> = Arc::new(PeerManager::new(
+		lightning_msg_handler,
+		keys_manager.get_node_secret(),
+		&ephemeral_bytes,
+		logger.clone(),
+	));
+
+	// ## Running LDK
+	// Step 16: Initialize Peer Connection Handling
+
+	// We poll for events in handle_ldk_events(..) rather than waiting for them over the
+	// mpsc::channel, so we can leave the event receiver as unused.
+	let (event_ntfn_sender, _event_ntfn_receiver) = mpsc::channel(2);
+	let peer_manager_connection_handler = peer_manager.clone();
+	let event_notifier = event_ntfn_sender.clone();
+	let listening_port = args.ldk_peer_listening_port;
+	tokio::spawn(async move {
+		let listener = std::net::TcpListener::bind(format!("0.0.0.0:{}", listening_port)).unwrap();
+		loop {
+			let tcp_stream = listener.accept().unwrap().0;
+			lightning_net_tokio::setup_inbound(
+				peer_manager_connection_handler.clone(),
+				event_notifier.clone(),
+				tcp_stream,
+			)
+			.await;
```

Comment:
I believe this waits until the connection is *closed* which is not what you want here.

Reply:
I think anything in a `tokio::spawn` makes progress regardless, since the threadpool implicitly awaits it.

Reply:
Sorry, that wasn't clear.  You would need another spawn on the `setup_inbound` to fire it off.